### PR TITLE
Improve collision detection for moving missiles

### LINKED ISSUE #1531: Blocking spells activates block animation and still deals damage
[single_0.zip](https://github.com/diasurgical/devilutionX/files/6313226/single_0.zip)

We tested using a single succubus enemy type shooting their Blood Stars. Sometimes the spell is blocked completely, but sometimes the spell gets blocked but still deals damage. The same issue is present in vanilla Hellfire.

Hellfire 1.0.1: [https://www.youtube.com/watch?v=8l9SDX6BPjQ](https://www.youtube.com/watch?v=8l9SDX6BPjQ) (thx Maxpire)
devilutionX 1.2.1 Hellfire: [https://cdn.discordapp.com/attachments/764970679564501052/831966291337543680/snow_witch_blocking.mp4](https://cdn.discordapp.com/attachments/764970679564501052/831966291337543680/snow_witch_blocking.mp4)

_In the second video you can see instances of "damage taken" sound triggering, but me not taking damage, that is most likely because I'm wearing 64% magic resistance and items with -5 damage taken._

The expected behaviour is that every blocked Blood Star should be fully blocked and deal no damage, regardless of damage reductions.

---

PS: Attached is my save from devilutionX 1.2.1 Hellfire in a position ready to test.

### LINKED ISSUE #2826: Arrows'/projectile spells' effective chance to hit influenced by projectile speed
**Important information**
Operating system / DevilutionX version / did you download a release or did you compile yourself?

https://github.com/diasurgical/devilutionX/commit/effee3f7bb7b3e3b868befb2006d3200e5e617a5

**Describe**
Slower missiles often perform a hit check multiple times as they spend time passing over a monster's tile, meaning that a projectile that should have low chance to hit gets multiple attempts to roll a hit and therefore has higher _effective_ chance to hit. This seems to be the more common case. However, faster missiles can sometimes skip a hit check altogether seemingly by never spending time on the same tile as a monster, meaning that faster missiles have lower _effective_ chance to hit, other things equal.

This can be tested by tracking the `hper` variable within the `MonsterMHit` method in missiles.cpp while running the game and shooting stuff with arrows and things like Firebolt. You can exaggerate the problem by altering the speed of arrows/firebolts to be extremely slow and by drastically reducing your (normal) hit chance, and more hit checks will be output even when firing single shots at one monster.

This seems like it might be a vanilla bug as well, but I'm unsure. Also possibly related to #1672 and #1531

**To Reproduce**
1. Get a character with bad chance to hit for magic or bows.
2. Shoot monsters.
3. Compare actual chance to successfully hit a monster with expected chance to hit.

**Expected behavior**
A projectile performs exactly one hit check against a monster it collides with. If it fails the first hit check, it continues past the initial monster and cannot attempt to hit it again. Faster projectiles should also always perform a hit check against a monster standing in the projectile's path.

**Screenshots**
sorctesting
hitchancedebug
I drastically lowered Sorcerer's hit chance in the code and shot some stuff on dlvl 1. The debug screenshot shows my chance to hit being returned on each hit check against a monster I shot at, and it also shows me tapping spacebar in between each shot. On the final shot against the Scavenger a couple tiles directly south of me, the game performed three hit checks as seen by the three 46s in a row, and ultimately landing the hit and killing it on the third hit check. This is with a normal speed Firebolt spell.

### LINKED ISSUE #2837: Arrows Bypass Target Tile at Certain Angles and Distances
**Important information**
Windows 10 / 1.2.1 / Downloaded Release

**Describe**
At certain angles and distances, your arrows will bypass the targeted tile. This can make it impossible to hit a monster without moving slightly to the side or another direction.

**To Reproduce**
1. Load attached save file.
2. Move mouse over the monster shown in the screenshot.
3. Click on it to your hearts content. Your arrows will never land.

**Expected behavior**
With no obstructions between the player and monster, the player should be able to hit them.

**Screenshots**
image

**Additional context**
If you move slightly to the right (non-isometrically, so thinking +1 in x-axis), the monster stays on screen and can be hit with arrows by clicking on them, and you can hit them pretty easily, so I do not believe this is a case of the to hit dropping to 0 from the distance.

[single_2.zip](https://github.com/diasurgical/devilutionX/files/7146605/single_2.zip)

### PR DESCRIPTION
**Pull request by obligaron**
Fixes #2826
Fixes #1531
Fixes #2837

Notes:

- First Commit is only refactoring as a base for the following commits
- Check if the missile skipped tiles and do also collision checking for these skipped tiles
- Prevent perform collision detection for a target more then once at a tile (that means you can also avoid a hit by missed hit chance and not only block)
- Tested with missiles 10x faster and 10x slower then normal

### CODE REVIEW

File: Source/missiles.cpp
Comment (ephphatha) [inline Source/missiles.cpp:614]: Given the way dMonster and dPlayer contain negated indexes for moving entities I think using bitwise combinations would be more straightforward than subtraction here.

XOR should be fine, could also shift the value in dMonster left 4-8 bits and mask the player index onto the low bits to avoid collisions (which is already pretty unlikely).
```suggestion
	int16_t tileTargetIdentifer = dMonster[missile.position.tile.x][missile.position.tile.y] ^ dPlayer[missile.position.tile.x][missile.position.tile.y];
```

File: Source/missiles.cpp
Comment (ephphatha) [inline Source/missiles.cpp:625]: This can be done by checking all the discrete tiles between prevTile and the current position, I'll have a go at simplifying the logic when I finish work.

(Been meaning to do the same for the animation logic as well, that's a case of working out the distance to the tile border as a proportion of the total movement in a frame then scaling the predicted movement by the result. That case can be done without a loop.)

**Review by ephphatha — COMMENTED** [review]
Combining missile movement and collision checking seems sensisble (especially given the observed bugs 😂 ). A few comments from a quick read through the combined function, I'll have a look at the call sites when I have more time this evening.

File: Source/missiles.cpp
Comment (StephenCWills) [reply Source/missiles.cpp:614]: Rather than unlikely, I think collisions are impossible. Can players and monsters occupy the same tile simultaneously?

Comment (StephenCWills) [reply Source/missiles.cpp:614]: If the indexes are negated for moving entities, doesn't that mean we might still check collisions twice for the same entity in the same tile if the entity starts or stops moving while the missile travels through that tile? Perhaps we should use `abs()`.

File: Source/missiles.cpp
Comment (StephenCWills) [inline Source/missiles.cpp:624]: If a missile passes over an adjacent tile into a diagonal tile in a single bound, this would skip the collision test for the adjacent tile.

File: Source/missiles.cpp
Comment (ephphatha) [reply Source/missiles.cpp:614]: > Rather than unlikely, I think collisions are impossible. Can players and monsters occupy the same tile simultaneously?
> 
> EDIT: I suppose maybe you meant hash collisions. 🤦

I did mean hash collisions :D, however hash collisions are only an issue if a player occupies the same tile as the golem they control, which I don't think is possible even if both entities are moving? Given the point you bring up in the second comment about hitting a moving entity multiple times using the first non-zero absolute value sounds fine.

Comment (StephenCWills) [reply Source/missiles.cpp:614]: > I did mean hash collisions :D, however hash collisions are only an issue if a player occupies the same tile as the golem they control, which I don't think is possible even if both entities are moving?

I have bad news.

https://user-images.githubusercontent.com/9203145/132784190-43ff7aec-6fd4-44ef-8e80-75481b2cb77f.mp4

Comment (StephenCWills) [reply Source/missiles.cpp:614]: While scanning through the DSF Buglist just now, I found this text on the subject of sharing tiles.

> **SHARING SQUARES**
> 
> When a character teleports/phases onto a square that a monster was moving in to, they will overlap. The character will have to move off the square to hit the monster. The monster, however, can attack the same square it's standing on.
> 
> If two characters wind up occupying the same square, which can happen rarely when casting Teleport or when Resurrecting while standing over a body, the player who does not move will become invisible to both players, until he/she moves. Further, the player first on the internal list of players may be the first targeted and attacked if a monster approaches.
> 
> The player can also share a square with a Golem.

Comment (NiteKat) [reply Source/missiles.cpp:614]: A player can also share a tile with a monster by spawning on that tile while a monster occupies it. The function for spawning players on a floor (such as by taking the stairs down) will put a player on a tile occupied by a monster happily. When this happens, the player cannot hit the monster until they move, but the monster is free to melee the player to death. Fun times. (if you attack the monster, your attack goes straight down on screen)

Comment (julealgon) [reply Source/missiles.cpp:614]: > I have bad news.

> A player can also share a tile with a monster by spawning on that tile while a monster occupies it.

I think it is a bad idea to base new logic on what are clearly existing bugs. Instead, we should fix the bugs and prevent the situation from ever happening.

Comment (StephenCWills) [reply Source/missiles.cpp:614]: > I think it is a bad idea to base new logic on what are clearly existing bugs. Instead, we should fix the bugs and prevent the situation from ever happening.

On that note, I think the hash collisions are pretty much a non-issue. Even in an imagined future state of the game where any monster could share an ID with any player, it would be extremely difficult to create a situation where the hash collision actually made a difference to this logic. Furthermore, the impact of the hash collision is negligible.

File: Source/missiles.cpp
Comment (obligaron) [reply Source/missiles.cpp]: Copied from the wrong location. 😁 
Fixed, thanks. 🙂

File: Source/missiles.h
Comment (obligaron) [reply Source/missiles.h]: Renamed it to `lastCollisionTargetHash` 🙂

File: Source/missiles.cpp
Comment (obligaron) [reply Source/missiles.cpp]: I'm also not sure if we should care much about it.
For me this are also bugs that should be fixed and @qndel was fast and fixed the first one already. 👍 
Do we have a issue for the other player spawning case?

But if wanted we could improve it now or in a later pr.
We could for example add two fields to `Missile`, one for player and one for monster.

Comment (NiteKat) [reply Source/missiles.cpp]: I don't think there's an issue for players and monsters occupying the same tiles. Though, now I'm curious about other ways to possibly do it.... if I get some time this weekend, I'll try to get some save files and confirm things and get some issues open.

File: Source/missiles.cpp
Comment (obligaron) [reply Source/missiles.cpp:625]: This would be great! Really looking forward to it. 🙂 
But I think the current version could be good enough to get this pr merged. Cause I dislike to have pr's too long open.

File: Source/missiles.cpp
Comment (obligaron) [reply Source/missiles.cpp]: Totally missed this one. 😄 
I tried to fix it with checking coordinates diff and only use `WalkingDistance` if the missile flies straight in one direction (only x or y).
Hope you can look again. 🙂

File: Source/missiles.cpp
Comment (StephenCWills) [inline Source/missiles.cpp:612]: Good thinking on checking if the missile is flying straight. I thought I'd mention that I saw the following function when I was reviewing this PR yesterday. It looks like you're using effectively the same logic in the ternary else clause.

https://github.com/diasurgical/devilutionX/blob/ec1850e2950f015d7814a8d9472bde471f4cb0d6/Source/engine/point.hpp#L142-L147

Comment (obligaron) [reply Source/missiles.cpp]: I also think it's the same, switched to `ManhattanDistance`.
This way it looks cleaner and less code. Thanks 🙂

File: Source/missiles.cpp
Comment (StephenCWills) [reply Source/missiles.cpp]: I still think that absolute value is a good idea, but the impact is similarly negligible so I'll approve the PR regardless.

**Review by StephenCWills — DISMISSED** [review]
I'm impressed by the speedy delivery of this PR. Since this does have an impact on game balance, I'd like to leave it to @AJenbo to decide whether to merge this now, but I approve of the changes regardless.

File: Source/missiles.cpp
Comment (obligaron) [reply Source/missiles.cpp]: I changed it to suggested `^`, it also doesn't hurt. 🙂

**Review by StephenCWills — APPROVED** [review]

File: Source/missiles.cpp
Comment (ephphatha) [reply Source/missiles.cpp:625]: Yep, I also got sidetracked yesterday and didn't get around to looking at it sorry 😅 . I'll look at both of these cases as a separate PR (though happy for someone else to take a look too).

Comment (StephenCWills) [reply Source/missiles.cpp:625]: TBH, I played around with the math a bit yesterday and found it to be rather difficult to implement. I was relieved when you said you would look into it, but I'd be happy to give it another try.

File: Source/loadsave.cpp
Comment (AJenbo) [inline Source/loadsave.cpp:688]: There are 3 bytes available from alignment, afaik they will always be zero for vanilla saves and it would allow us to retain the sate past saves.

File: Source/missiles.cpp
Comment (AJenbo) [inline Source/missiles.cpp:614]: I prefer lines to be 80, and hard limited to 120 chars. This one is 212 which just barley fits on half of my ultra wide monitor in this differ with 100% zoom.

```suggestion
	Point position = missile.position.tile;
	int possibleVisitTiles = (missile.position.velocity.deltaX == 0 || missile.position.velocity.deltaY == 0) ? prevTile.WalkingDistance(position) : prevTile.ManhattanDistance(position);

	int16_t tileTargetHash = dMonster[position.x][position.y] ^ dPlayer[position.x][position.y];
```
Even with this change it's still 186, but that's a bit better. We can wait with further reducing till it's gets added as a clang rule :)

File: Source/missiles.cpp
Comment (StephenCWills) [reply Source/missiles.cpp:625]: @ephphatha My current line of thinking is rather than working out the distance to the edge of each tile, it would probably be a lot easier to calculate the points of collision with horizontal/vertical lines that cut through the centers of the tiles. This way, you could simply iterate through the horizontal/vertical lines in the missile's path and handle corners as a special case. Plus, you shouldn't have to deal with the possibility of truncation errors putting the missile in the wrong tile during calculations like you would when straddling the edge.

Since I'm struggling with the implementation, I figured I'd share this idea in case the information helps get you there faster than me.

image

**Review by AJenbo — DISMISSED** [review]
Definitely something we should merge, it solves most of the bugs caused by the dodgy collision code.

I do slightly wish that there was less "will work if x never happens", but I don't feel they are unreasonable assumptions and the alternative would probably be some major reworkings.

Since there are a few active debates still going, I will not merge it right away but anyone is free to do so if you feel it makes more sense to revise with additional PRs.

File: Source/missiles.cpp
Comment (obligaron) [reply Source/missiles.cpp:614]: If you want I can change it to this:
```cpp
	int possibleVisitTiles;
	if (missile.position.velocity.deltaX == 0 || missile.position.velocity.deltaY == 0)
		possibleVisitTiles = prevTile.WalkingDistance(missile.position.tile);
	else
		possibleVisitTiles = prevTile.ManhattanDistance(missile.position.tile);
```
This has max 87 chars and avoids a new local variable... but now we get 4 instead of 1 line. 😉

File: Source/loadsave.cpp
Comment (obligaron) [reply Source/loadsave.cpp:688]: I'm not sure if we should use it for this atm, cause
- DevilutionX (includes Version 1.2.1) writes uninitialized data to this (was changed in #2287). So I think it would be good to have version 1.3 out so that most saves should contain zero initialized values (saved in vanilla or devilutionx 1.3).
- The effect is not that great. At maximum you get hit twice once.
- I feel like this bytes are precious and perhaps we could use them for something even better. :innocent:

File: Source/missiles.cpp
Comment (qndel) [reply Source/missiles.cpp:614]: funny, I wanted to add manhattan distance recently :D

Comment (AJenbo) [reply Source/missiles.cpp:614]: To me long lines are worse then many lines. They are very hard to see structure in and if line length is not standardized eh complexity of two function with the same line count can be vastly different.

I'm not a big fan of uninitialized variables, but seeing this I think it's the best of the 3 options.

Comment (AJenbo) [reply Source/missiles.cpp:614]: > funny, I wanted to add manhattan distance recently :D

You are 3 months late :)

Comment (obligaron) [reply Source/missiles.cpp]: Changed it. 🙂

**Comment by Chance4us** [comment]
I haven't tested it for long, but the first impression was that you hit the opponents much easier. This greatly reduces the difficulty of the game. You could of course do some balancing by making the opponents stronger in a way, but it's all a balancing act between the bugs by design and the feeling of originality. In my opinion, we should therefore consider making the change unlockable via an optional switch.

**Comment by StephenCWills** [comment]
> I haven't tested it for long, but the first impression was that you hit the opponents much easier. This greatly reduces the difficulty of the game.

If you are using arrows or high level spells, that would make sense. Arrows move at a speed very close to 1 tile per tick so you'd be much more likely to miss a tile that the projectile passes through than get multiple collisions in the same tile.

> You could of course do some balancing by making the opponents stronger in a way, but it's all a balancing act between the bugs by design and the feeling of originality.

I originally thought we'd need to adjust the hit chance of missiles, but I've changed my mind precisely because arrow speed is so close to one tile per tick. I'm still a little concerned that this could adversely affect low-level mages which are notoriously difficult already, but otherwise this just feels like a nasty bug.

> In my opinion, we should therefore consider making the change unlockable via an optional switch.

I assume you mean to say that the hit chance adjustment could be an optional change, not this bug. In that case, I'd also say it's worth considering.

**Comment by Chance4us** [comment]
The hit chance of the arrows should be reduced and the hit chance of the rocket should be increased (see # 2890),
otherwise we are dealing with a heavily modified gameplay, which does not even come close to the original.
The way it feels now is a different game.

**Comment by Chance4us** [comment]
@StephenCWills I don't mean to say that the changes are fundamentally wrong or bad.
I don't mean that in a negative way and such changes are of course welcome. 
After all, they fix bugs that were not considered or overlooked during development.
Since these actually feel a lot different from the original, I recommend combining them into one option.
Or the other way around, to summarize the vanilla peculiarities and diseases in one option and to award with the title "hit chance classic mode" for example.

**Comment by StephenCWills** [comment]
> The hit chance of the arrows should be reduced and the hit chance of the missiles should be increased (see #2890),
> otherwise we are dealing with a heavily modified gameplay mechanics, which does not even come close to the original.
> The way it feels now is an another different game that wants to be perfect.

I recommend you review #2826 to gain a better understanding of what effect this PR has on projectiles. If you insist that the accuracy of arrows needs to be reduced, you're going to have to convince me with data. Spells are much trickier because the projectile speed is affected by spell level. I can appreciate that much of your work here is based on feel, and I don't want to discount that. It's just that the solution can't be as simple as "turn these two knobs and we have the original game again".

**Comment by Chance4us** [comment]
I understand that. I see that from the player's side. And I'm not a good programmer at all. From here on we're slowly moving into a gray area, where other players may also say, wait a minute, the original was somehow different. What do we want to achieve with Devilutionx, a mod or the original just with more comfort and fewer annoying bugs? The players will think that we will no longer hit the enemy in the same way as before and also the other way around, my hero will be hit differently or, strangely enough, not hit.

I can't convince you with data or numbers,
but I can roughly describe the problem with the arrows.
You have to imagine you are in level 5 and surrounded by the monsters (Goat Man https://diablo-archive.fandom.com/wiki/Goat_Man_(Diablo_I)). These monsters love to run around, criss-cross, orbit the hero until they attack him directly. While they are circling the hero they are as good as impossible to hit, especially the most difficult to hit diagonally. That makes the game very tough and difficult in this situation.
The current change makes it possible to hit opponents more easily at any angle, and it makes the game feel less difficult. 

With the missiles on the other hand, I felt the opposite effect and demonstrated it with the help of the numbers. Here we are dealing with a reduced hit rate. As a player, I find the bad hit rate with the missiles worse than the problem with the arrows, because the missile flies through for no good reason and even though you aim very precisely. I don't know how but we should find a compromise, because otherwise the game feels either too easy or too difficult depending on the situation.

**Comment by julealgon** [comment]
> While they are circling the hero they are as good as impossible to hit, especially the most difficult to hit diagonally. That makes the game very tough and difficult in this situation.

That was because the hit checks were completely wrong.

> The current change makes it possible to hit opponents more easily at any angle, and it makes the game feel less difficult.

There were other changes that impact gameplay as well, for example the removal of horizontal walking, which will make the game harder. Slower spells will also hit less, which makes the game harder.

Additionally, missile logic not only impacts players, but monsters too. So monsters will hit you with projectiles more often when you are moving, and less often if they fire very slow missiles (due to previous multiple hit checks).

I don't think we should try to rebalance the game. We are here to fix bugs. If someone wants a rebalance after the bugs are fixed, that should be done in a mod.

**Comment by AJenbo** [comment]
> What do we want to achieve with Devilutionx, a mod or the original just with more comfort and fewer annoying bugs?

Getting hit 11 times by the same missile is very annoying, also arrows skipping some tiles making it impossible to hit is also very annoying. I do appreciate that fixing issues like this will affect game balance and as such it is something we have to carefully evaluate additionally. But I think we already started down this road by fixing the AI not being able to locate and avoid firewalls as well as firewalls dealing damage based on dungeon level instead of hero level.

**Pull request merged**
